design_notes/XDN03 - Platform Support for Activation Factories.md

```diff
@@ -0,0 +1,76 @@
+
+## Overview
+
+As described in the Type System Specification, a class that is activatable, composable, and/or has static methods requires an activation factory.
+
+When consuming code needs an instance of a factory, it should not require knowledge of which component implements that factory, nor should it be responsible for dynamically loading that module and searching for exports.
+
+For that reason, consuming code will request factories through the PAL.
+The PAL will consult its mapping of class names to component libraries, load (if neccessary) the library, and then call the exported method on the library, returning the result to the caller.
+
+This implies three pieces of functionality that work together to support activation:
+* A function exported by the PAL for client code to request factories.
+* A function exported by components to retrieve instances of its factories.
+* A scheme for the PAL to build a mapping of class names to component libraries.
+
+## Exporting factories from components
+
+For the PAL to retrieve factories, xlang components must implement and export a function, **xlang_lib_get_activation_factory**, which accepts a class name in string form and the GUID of the requested interface, and returns the newly created class factory.
+Components providing factories are expected to correctly handle calls to this function from multiple threads concurrently.
+
+For components providing class factories, it is risky to have factories with state that depends on other components.
+Because factories are requested and initialized in response to requests to construct a runtime class, a dependency cycle between components' factories could easily lead to their initialization methods being called reentrantly.
+This problem is easily avoided, and correct multithreaded behavior is simpler, if class factories are stateless.
+
+## Retrieving factory interfaces
+
+The PAL provides the function **xlang_get_activation_factory**, which accepts as input the class name in string form, and the GUID of the requested interface.
+Upon success, the output is the requested interface.
+
+This function may fail if:
+* The PAL can not map the class name to a component.
+* The PAL can not load the mapped component.
+* The PAL loads the component, but can not locate the exported **xlang_lib_get_activation_factory** function to retrieve the component's factories.
+* The component itself fails to return the requested factory.
+
+## Mapping classes to components
+
+This functionality is expected to evolve as xlang's application model matures, with new behaviors and abstractions being added as appropriate.
+
+Upon the first request for an activation factory, or in other words, the first time a class name is passed to **xlang_get_activation_factory**, the PAL will attempt to load the owning component in the following manner:
+* Obtain the innermost namespace scope from the class name
```

Comment:
This is basically the same algorithm windows uses to locate metadata at runtime in winmd files.